File: python/1-50/18. 4Sum.py
class Solution(object):
    def fourSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        nums.sort()
        res = []
        for i in range(len(nums)):
            for j in range(i + 1, len(nums)):
                temp = target - nums[i] - nums[j]
                l, r = j + 1, len(nums) - 1
                while l < r:
                    if nums[l] + nums[r] > temp:
                        r -= 1
                    elif nums[l] + nums[r] < temp:
                        l += 1
                    else:
                        if [nums[i], nums[j], nums[l], nums[r]] not in res:
                            res.append([nums[i], nums[j], nums[l], nums[r]])
                        l += 1
        return res

File: cpp/1-50/18. 4Sum.cpp
class Solution
{
public:
    vector<vector<int> > fourSum(vector<int>& nums, int target)
    {
        vector<vector<int> > res;
        sort(nums.begin(), nums.end());
        if(nums.size() < 4)
            return res;
        for(int i = 0; i < nums.size() - 3;)
        {
            for(int j = i+1; j < nums.size() - 2;)
            {
                int low = j + 1;
                int high = nums.size() - 1;
                while(low < high)
                {
                    int temp = nums[i] + nums[j] + nums[low] + nums[high];
                    if(temp == target){
                        vector<int> temp({nums[i], nums[j], nums[low], nums[high]});
                        res.push_back(temp);
                        do{
                            low++;
                        }while(low < high && nums[low] == nums[low-1]);
                        do{
                            high--;
                        }while(low < high && nums[high] == nums[high + 1]);
                    }
                    if(temp < target)
                    {
                        do{
                            low++;
                        }while(low < high && nums[low] == nums[low-1]);

                    }
                    if(temp > target)
                    {
                        do{
                            high--;
                        }while(low < high && nums[high] == nums[high + 1]);
                    }
                }
                do{
                    j++;
                }while(j < nums.size() - 2 && nums[j] == nums[j-1]);
            }
            do{
                i++;
            }while(i < nums.size() - 3 && nums[i] == nums[i-1]);
        }
        return res;
    }
};